### ISSUE TITLE
Awesome idea 👏🏼

### ISSUE THREAD

**Issue by parro-it** [mischnic/screenshot-tester#1]
I was thinking we need the same tool from same days.

I remember many time ago I read an article about the same technique used to test a CSS framework.
I don't remember which one unfortunately, but I remember the difference between the stored screenshot and the testing one was complicated to do correctly...

It try to avoid false negative by don't taking into account small difference. But maybe this is not necessary for native GUI apps?

**Comment by mischnic** [comment]
The main problem I was having on Windows is that sometimes the window is focused and sometimes it's not.

The `tolerance` here mitigates that to a certain extent:

https://github.com/mischnic/screenshot-tester/blob/0d735f4107bb2268583f760a36bfa7a2e1ec95c5/index.js#L60


Do you know a somewhat cross-distribution way/program on Linux to take a screenshot of a specific window?

**Comment by mischnic** [comment]
Oh, and two libui-node examples aren't suited for this: the core-api and control-gallery show datetime pickers, who's value on Windows is the current time.

**Comment by parro-it** [comment]
> Oh, and two libui-node examples aren't suited for this: the core-api and control-gallery show datetime pickers, who's value on Windows is the current time.

Yes actually they doesn't have setter and getters, but there is a PR on libui by @cody271: https://github.com/andlabs/libui/pull/297.

As a workaround, we can change system time to a fixed date before running the tests, and restore it after it run.

**Comment by parro-it** [comment]
> Do you know a somewhat cross-distribution way/program on Linux to take a screenshot of a specific window?

I think robotjs could be a good fit: https://robotjs.io/docs/syntax#screencapturex-y-width-height

**Comment by parro-it** [comment]
If you make the executable customizable, the tool could be used also to test libui directly:  

https://github.com/mischnic/screenshot-tester/blob/0d735f4107bb2268583f760a36bfa7a2e1ec95c5/index.js#L50-L55

What do you think about?

**Comment by parro-it** [comment]
> The main issue I was having on Windows is that sometimes the window is focused and sometimes it's not.

https://github.com/sindresorhus/active-win could be of help.

**Comment by parro-it** [comment]
There is an open issue on `active-win` to purpose to trigger an event when a window get focus:
https://github.com/sindresorhus/active-win/issues/1

**Comment by mischnic** [comment]
> If you make the executable customizable, the tool could be used also to test libui directly:

Done

**Comment by mischnic** [comment]
> > The main issue I was having on Windows is that sometimes the window is focused and sometimes it's not.
>
> https://github.com/sindresorhus/active-win could be of help.

> There is an open issue on active-win to purpose to trigger an event when a window get focus:
sindresorhus/active-win#1


It works fine though if you don't touch anything during testing (because the windows will always be in focus after opening).

**Comment by mischnic** [comment]
> I don't know how different machines of the same OS would compare.

I'll have to tweak the code a bit. The reference images on the left were made on my machine, the right ones by Travis: https://snapshots-lagwwdfhwj.now.sh/

**Comment by mischnic** [comment]
![apr -30-2018 20-26-27](https://user-images.githubusercontent.com/4586894/39443413-d2548518-4cb4-11e8-8b91-2f63168a0237.gif)

😕

**Comment by mischnic** [comment]
Apart from that, it works quite well: https://snapshots-pklqmkerlv.now.sh/

**Comment by parro-it** [comment]
Good work! How is you travis build configured? 
Maybe the difference depends on some different configuration of screen colors/resolution?

**Comment by mischnic** [comment]
> Maybe the difference depends on some different configuration of screen colors/resolution?

No change.

Could you try it out on your machine? Clone, `yarn` or `npm install` in the root and example folder and then `node example/index.js` 

If something is missing: https://github.com/mischnic/screenshot-tester/blob/ci-test/.travis.yml

**Comment by parro-it** [comment]
Sure! Example folder has no package.json. Should I work in ci-test branch?

**Comment by mischnic** [comment]
Yes, forgot about that.

**Comment by parro-it** [comment]
It seems I lack a python click module.
Could I install it with `pip install click` or is it from another package?

**Comment by mischnic** [comment]
Yes, you might also need another package (pip install click pyobjc-framework-Quartz)

**Comment by parro-it** [comment]
Output:

Failed: area-adv.js - "libui textDrawArea Example", see snapshots/temp/area-adv.js_diff.png
Failed: text.js - "libui textDrawArea Example", see snapshots/temp/text.js_diff.png
Failed: forms.js - "Forms window", see snapshots/temp/forms.js_diff.png
Failed: grid.js - "Forms window", see snapshots/temp/grid.js_diff.png
Failed: node-pad.js - "Node Pad", see snapshots/temp/node-pad.js_diff.png
Generated HTML report: snapshots/index.html

Where could I upload the HTML?
As a preview, the gradients of area-adv.js (both the circle and the square) are signed as different.
the first row of text has the same background color of the emoji on my machine.
The other examples has only some little difference on fonts.

I'm running on High Sierra.

Evaluating with a human brain, I would say we already found a bug on text example. The other difference should have been skipped as minor ones.

**Comment by mischnic** [comment]
I used https://zeit.co/blog/unlimited-static. 

Install `now` globally with npm and run it in the snapshots folder (you might have to create an account). You'll get an url like the ones I posted previously.

**Comment by parro-it** [comment]
Done: https://snapshots-wxultoqlxo.now.sh/

**Comment by mischnic** [comment]
Mille grazie!

**Comment by parro-it** [comment]
Non c'è di che! 😆
With a tolerance of 60 it reflect what I think is a correct result:
```
Passed: area-adv.js - "libui textDrawArea Example"
Failed: text.js - "libui textDrawArea Example", see snapshots/temp/text.js_diff.png
Passed: forms.js - "Forms window"
Passed: grid.js - "Forms window"
Passed: node-pad.js - "Node Pad"
Generated HTML report: snapshots/index.html
```
https://snapshots-pjadazqovu.now.sh/

**Comment by mischnic** [comment]
With a tolerance of 5, white is the same as the grey background: (the added combobox)
I'll have to use a different comparison algorithm (#4)


![text js_diff](https://user-images.githubusercontent.com/4586894/39467549-e90d58e6-4d2e-11e8-99e6-df3e609582ea.png)
![text js](https://user-images.githubusercontent.com/4586894/39467550-e92ec54e-4d2e-11e8-9d59-b1c29881b7c0.png)

**Comment by mischnic** [comment]
Could you please test it again? I switched to a different diffing library

**Comment by parro-it** [comment]
I'm commuting to home, I will test within some hour...

**Comment by parro-it** [comment]
Much better now!

![image](https://user-images.githubusercontent.com/11197111/39539018-e065de46-4e3e-11e8-899a-da6bf78d0be5.png)

https://snapshots-mtbpfureqh.now.sh

Good work!

**Comment by mischnic** [comment]
What did I do wrong with the appveyor config? https://ci.appveyor.com/project/mischnic/screenshot-tester/build/1.0.1/job/94rvxg7b6bte2x3o
```
UiDrawPath.obj : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\projects\screenshot-tester\example\node_modules\libui-node\build\nbind.vcxproj]
```

**Comment by parro-it** [comment]
Uhm... could you try with npm? 
Otherwise, it seems there is a conflict name with the DEBUG marco...

**Comment by parro-it** [comment]
that's strange... but had you done that intentionally?

**Comment by mischnic** [comment]
No 😄. It should do the same as travis.

**Comment by mischnic** [comment]
Fixed. Now I know what `build: off` does in the config...

**Comment by parro-it** [comment]
I didn't note that difference too, 
I was comparing this and libui-node appveyor config but they seems the same 🤓
